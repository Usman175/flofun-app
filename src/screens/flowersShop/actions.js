import { fetchFlowers } from '../../../constants/api'

export const FETCH_FLOWERS = 'FETCH_FLOWERS'
export const SELECT_FLOWER = 'SELECT_FLOWER'
export const REMOVE_FLOWER = 'REMOVE_FLOWER'
export const CHANGE_MESSAGE = 'CHANGE_MESSAGE'

export const fetchAvailableFlowers = () => ({
  type: FETCH_FLOWERS,
  payload: fetchFlowers(),
})

export const selectFlower = flower => ({
  type: SELECT_FLOWER,
  payload: flower,
})

export const removeSelectedFlower = flower => ({
  type: REMOVE_FLOWER,
  payload: flower,
})

export const handleChangeMessage = (message, key) => ({
  type: CHANGE_MESSAGE,
  payload: { message, key },
})
