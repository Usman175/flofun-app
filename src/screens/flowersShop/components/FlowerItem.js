import React, { Component } from 'react'
import {
  View,
  Text,
  Image,
  TouchableOpacity,
  TouchableWithoutFeedback,
  Animated,
} from 'react-native'
import { MaterialIcons } from '@expo/vector-icons'

import styles from './styles/FlowerItem'

export default class FlowerItem extends Component {
  constructor(props) {
    super(props)

    this._animatedWidth = new Animated.Value(0)
    this._animatedQuantity = new Animated.Value(0)
    this.state = {
      isOpen: false,
      quantity: 1,
    }
  }

  closeQuantitySelector = () => {
    this.setState({ isOpen: false })
    Animated.spring(this._animatedWidth, {
      toValue: 0,
    }).start()
  }

  changeQuantity = change => {
    const quantity = this.state.quantity + change

    if (quantity === 0) {
      this.closeQuantitySelector()
    } else {
      Animated.sequence([
        Animated.timing(this._animatedQuantity, {
          toValue: -3,
          duration: 100,
        }),
        Animated.timing(this._animatedQuantity, {
          toValue: 3,
          duration: 100,
        }),
        Animated.timing(this._animatedQuantity, {
          toValue: 0,
          duration: 100,
        }),
      ]).start()
      this.setState({ quantity })
    }
  }

  handleSelectFlower = flower => {
    if (this.state.isOpen) {
      this.closeQuantitySelector()
      this.props.selectFlower({ ...flower, quantity: this.state.quantity })
    } else {
      this.setState({ isOpen: true })
      Animated.spring(this._animatedWidth, {
        toValue: 1,
      }).start()
    }
  }

  renderQuantitySelector = () => {
    const { flower, selected } = this.props
    const quantity = this.state.quantity

    const containerStyles = [
      styles.quantitySection,
      {
        width: this._animatedWidth.interpolate({
          inputRange: [0, 1],
          outputRange: ['0%', '55%'],
          extrapolate: 'clamp',
        }),
        transform: [{ scale: this._animatedWidth }],
      },
    ]

    const titleContainerStyles = [
      styles.titleSection,
      {
        width: this._animatedWidth.interpolate({
          inputRange: [0, 1],
          outputRange: ['55%', '0%'],
          extrapolate: 'clamp',
        }),
        paddingLeft: this._animatedWidth.interpolate({
          inputRange: [0, 1],
          outputRange: [5, 0],
          extrapolate: 'clamp',
        }),
      },
    ]

    const quantityStyles = [
      styles.quantityText,
      {
        transform: [
          {
            translateX: this._animatedQuantity,
          },
        ],
      },
    ]

    return (
      <View style={styles.flowerDesc}>
        <Animated.View style={titleContainerStyles}>
          <Text ellipsizeMode="tail" numberOfLines={2} style={styles.title}>{flower.title}</Text>
        </Animated.View>
        <View style={[styles.priceSection, selected ? styles.priceSelected : null]}>
          <Text style={styles.price}>$ {flower.price}</Text>
        </View>
        <Animated.View style={containerStyles}>
          <TouchableOpacity
            style={styles.quantitySelectorButton}
            onPress={() => this.changeQuantity(-1)}
          >
            <Text style={styles.quantitySelectorButtonText}>-</Text>
          </TouchableOpacity>
          <Animated.Text style={quantityStyles}>
            {quantity}
          </Animated.Text>
          <TouchableOpacity
            style={styles.quantitySelectorButton}
            onPress={() => this.changeQuantity(1)}
          >
            <Text style={styles.quantitySelectorButtonText}>+</Text>
          </TouchableOpacity>
        </Animated.View>
        <View
          style={[styles.addToCart, this.state.isOpen || selected ? styles.priceSelected : null]}
        >
          <TouchableOpacity onPress={() => this.handleSelectFlower(flower)}>
            {selected
              ? <MaterialIcons name="check-circle" size={30} color="#fff" />
              : <MaterialIcons name="add-shopping-cart" size={30} color="#fff" />}
          </TouchableOpacity>
        </View>
      </View>
    )
  }

  render() {
    const { flower, lastElement } = this.props

    return (
      <View style={[styles.root, lastElement && styles.lastFlower]}>
        <View style={styles.imageSection}>
          <TouchableWithoutFeedback onPress={this._openInImageGallery}>
            <Image source={{ uri: flower.image }} style={styles.image} />
          </TouchableWithoutFeedback>
        </View>
        {this.renderQuantitySelector()}
      </View>
    )
  }
}
