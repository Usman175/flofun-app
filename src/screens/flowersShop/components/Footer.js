import React, { Component } from 'react'
import { View, TouchableOpacity, Image, TouchableWithoutFeedback, Animated } from 'react-native'
import { FontAwesome, MaterialCommunityIcons } from '@expo/vector-icons'

import styles from './styles/Footer'

export default class Footer extends Component {
  componentWillUpdate() {
    this.position = new Animated.ValueXY({ x: 0, y: 80 })

    Animated.spring(this.position, {
      toValue: { x: 0, y: 0 },
      duration: 500,
    }).start()
  }

  render() {
    const { selectedFlowers, removeSelectedFlower, navigate } = this.props
    return (
      <View style={styles.bottomContainer}>
        <View style={styles.bottomLeft}>
          {selectedFlowers.map((flower, key) => {
            const lastElement = key === selectedFlowers.length - 1

            if (lastElement) {
              return (
                <Animated.View key={flower._id} style={[this.position.getLayout(), styles.circle]}>
                  <Image source={{ uri: flower.image }} style={styles.selectedImage} />
                  <TouchableWithoutFeedback onPress={() => removeSelectedFlower(flower)}>
                    <MaterialCommunityIcons
                      style={styles.closeIcon}
                      name="close-circle"
                      size={20}
                      color="#000"
                    />
                  </TouchableWithoutFeedback>
                </Animated.View>
              )
            }

            return (
              <View key={flower._id} style={styles.circle}>
                <Image source={{ uri: flower.image }} style={styles.selectedImage} />
                <TouchableWithoutFeedback onPress={() => removeSelectedFlower(flower)}>
                  <MaterialCommunityIcons
                    style={styles.closeIcon}
                    name="close-circle"
                    size={20}
                    color="#000"
                  />
                </TouchableWithoutFeedback>
              </View>
            )
          })}
        </View>
        <View style={styles.bottomRight}>
          <TouchableOpacity onPress={() => navigate('ContactDetails')}>
            <FontAwesome name="arrow-right" size={40} color="#fff" />
          </TouchableOpacity>
        </View>
      </View>
    )
  }
}
