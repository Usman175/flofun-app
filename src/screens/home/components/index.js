import FlowerList from './FlowerList'

export {
  FlowerList
}
