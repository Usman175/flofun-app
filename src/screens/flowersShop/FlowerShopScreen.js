import React, { Component } from 'react'
import { View, Text, StatusBar } from 'react-native'
import { connect } from 'react-redux'
import ImageGallery from '@expo/react-native-image-gallery'

import { FlowerList, Footer } from './components'
import LoadingScreen from '../../commons/LoadingScreen'
import styles from './styles/FlowerShopScreen'
import HeaderStack from '../../commons/HeaderStack'
import {
  fetchAvailableFlowers,
  selectFlower,
  removeSelectedFlower,
} from './actions'

@connect(
  state => ({
    flowers: state.flowerShop.flowers,
    selectedFlowers: state.flowerShop.selectedFlowers,
  }),
  {
    fetchAvailableFlowers,
    selectFlower,
    removeSelectedFlower,
  }
)
class FlowerShopScreen extends Component {
  static navigationOptions = ({ navigation }) => ({
    title: 'Select Flower',
    ...HeaderStack(navigation.goBack),
  })

  componentDidMount() {
    this.props.fetchAvailableFlowers()
  }

  render() {
    const {
      flowers: {
        data,
        isFetched,
        error,
      },
      selectedFlowers,
    } = this.props

    if (!isFetched) {
      return <LoadingScreen />
    } else if (error.on) {
      return (
        <View>
          <Text>{error.message}</Text>
        </View>
      )
    }

    return (
      <View style={styles.root}>
        <StatusBar barStyle="light-content" />
        <View style={styles.contentContainer}>
          <FlowerList
            flowers={data}
            selectFlower={this.props.selectFlower}
            selectedFlowers={selectedFlowers}
          />
        </View>
        <Footer
          selectedFlowers={selectedFlowers}
          removeSelectedFlower={this.props.removeSelectedFlower}
          navigate={this.props.navigation.navigate}
        />
        <ImageGallery />
      </View>
    )
  }
}

export default FlowerShopScreen
