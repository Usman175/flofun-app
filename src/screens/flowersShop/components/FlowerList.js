import React from 'react'
import { View, ScrollView } from 'react-native'
import styles from './styles/FlowerList'
import FlowerItem from './FlowerItem'

const FlowerList = ({ flowers, selectFlower, selectedFlowers }) =>
  <View style={styles.root}>
    <View style={styles.contentContainer}>
      <ScrollView vertical>
        {flowers.map((flower, key) =>
          <View key={flower._id} style={styles.flowerCard}>
            <FlowerItem
              flower={flower}
              selectFlower={selectFlower}
              selected={!!selectedFlowers.find(f => f._id === flower._id)}
              lastElement={key === flowers.length - 1}
            />
          </View>,
        )}
      </ScrollView>
    </View>
  </View>

export default FlowerList
