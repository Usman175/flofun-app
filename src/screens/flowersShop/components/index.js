import FlowerList from './FlowerList'
import Footer from './Footer'

export { FlowerList, Footer }
