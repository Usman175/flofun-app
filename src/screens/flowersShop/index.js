import FlowerShopScreen from './FlowerShopScreen'
import FlowerShopReducer from './reducer'

export {
  FlowerShopScreen,
  FlowerShopReducer,
}
