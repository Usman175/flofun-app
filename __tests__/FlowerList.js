import 'react-native'
import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import EStyleSheet from 'react-native-extended-stylesheet'
import Colors from '../constants/colors'
import { FlowerList } from '../src/screens/flowersShop/components/'

beforeAll(() => {
  EStyleSheet.build(Colors)
})

const img = 'https://cdn.arstechnica.net/wp-content/uploads/2016/02/5718897981_10faa45ac3_b-640x624.jpg'
const flowers = [
  { title: 'flower 1', image: img, price: 10.90 },
  { title: 'flower 2', image: img, price: 12.50 },
  { title: 'flower 3', image: img, price: 14.20 },
]

describe('FlowerList', () => {
  test('renders correctly', () => {
    const flowersElement = shallow(
      <FlowerList
        flowers={flowers}
        selectedFlowers={[]}
      />
    )

    expect(toJson(flowersElement)).toMatchSnapshot()
  })
})
