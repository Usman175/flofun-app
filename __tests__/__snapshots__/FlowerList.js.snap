// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlowerList renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <ScrollViewMock
      vertical={true}
    >
      <View
        style={Object {}}
      >
        <FlowerItem
          flower={
            Object {
              "image": "https://cdn.arstechnica.net/wp-content/uploads/2016/02/5718897981_10faa45ac3_b-640x624.jpg",
              "price": 10.9,
              "title": "flower 1",
            }
          }
          selected={false}
        />
      </View>
      <View
        style={Object {}}
      >
        <FlowerItem
          flower={
            Object {
              "image": "https://cdn.arstechnica.net/wp-content/uploads/2016/02/5718897981_10faa45ac3_b-640x624.jpg",
              "price": 12.5,
              "title": "flower 2",
            }
          }
          selected={false}
        />
      </View>
      <View
        style={Object {}}
      >
        <FlowerItem
          flower={
            Object {
              "image": "https://cdn.arstechnica.net/wp-content/uploads/2016/02/5718897981_10faa45ac3_b-640x624.jpg",
              "price": 14.2,
              "title": "flower 3",
            }
          }
          selected={false}
        />
      </View>
    </ScrollViewMock>
  </View>
</View>
`;
